### PR TITLE
Optimize Tasks collections in .Net 9

### PR DESCRIPTION
**Pull request by LukaszRozmej**
Builds upon  #7532

## Changes

- Move task collections to pooled ones and use spans when using `WaitAll`, `WhenAll` and `WhenAny`

#### What types of changes does your code introduce?

- [ ] Bugfix (a non-breaking change that fixes an issue)
- [ ] New feature (a non-breaking change that adds functionality)
- [ ] Breaking change (a change that causes existing functionality not to work as expected)
- [x] Optimization
- [x] Refactoring
- [ ] Documentation update
- [ ] Build-related changes
- [ ] Other: _Description_

#### Requires testing

- [x] Yes
- [ ] No

#### If yes, did you write tests?

- [ ] Yes
- [x] No

### CODE REVIEW

File: src/Nethermind/Nethermind.Consensus/Producers/MultipleBlockProducer.cs [hunk 1]
```diff
@@ -32,23 +33,20 @@ protected MultipleBlockProducer(
...
+                    ? Task.FromResult<Block?>(null)
+                    : blockProducerInfo.BlockProducer.BuildBlock(parentHeader, blockProducerInfo.BlockTracer, cancellationToken: token));
             }
 
             IEnumerable<(Block? Block, T BlockProducer)> blocksWithProducers;
 
             try
             {
-                Block?[] blocks = await Task.WhenAll(produceTasks);
+                Block?[] blocks = await Task.WhenAll<Block?>(produceTasks.AsSpan());
```
Comment (rubo) [inline src/Nethermind/Nethermind.Consensus/Producers/MultipleBlockProducer.cs:49]: Do we really need `.AsSpan()`?

**Review by rubo — APPROVED** [review]
Nitpick: In some places we add items to the pool in a `for` loop with `.Add()`, in other places, with LINQ's `.Select()` or `Enumerable.Range().Select()`. Isn't it worth unifying the approach?

**Review by benaadams — COMMENTED** [review]
Why the .orig files?

**Comment by LukaszRozmej** [comment]
> Why the .orig files?

github merge fail

**Review by benaadams — APPROVED** [review]

**Pull request merged**